docs/proposals/20191016-e2e-test-framework.md

```diff
@@ -0,0 +1,199 @@
+
+By writing down the expected behaviors of Cluster API in the form of a test framework, we will be able to know if our changes break those expected behaviors and thus satisfy this user story.
+
+#### As a Cluster API developer, I want to be sure I’m not accidentally disrupting existing providers.
+
+Existing providers will be depending on Cluster API behavior. This is an alternative perspective of the previous user story.
+
+#### As the release manager of a Cluster API provider, I want to be able to recommend new releases to users as soon as they come up
+
+Not all providers have automated testing that use all components of Cluster API and test interactions. This framework would be able to provide that signal to release managers to know if a release will at least pass this suite.
+
+#### As a user of Cluster API, I want to be sure new releases I’m installing won’t cause regressions.
+
+Regressions are easy to introduce. If we happen to introduce a regression that is not covered by these behaviors we will be able to identify how to write a test to reproduce the bad behavior and to fix the underlying bug.
+
+### Implementation Details/Notes/Constraints
+
+This framework will be implemented using the Ginkgo/Gomega behavioral testing framework because that is the Kubernetes standard and does a good job of structuring and organizing behavior based tests.
+
+This framework will live in the `sigs.k8s.io/cluster-api` module because it will be formalizing behaviors of Cluster API. Therefore the set of behaviors for each version may change, so will the test suite.
+
+This framework is not conformance nor will be built in the same style as the conformance framework of kubernetes. That is to say, there will not be a distributed binary. The only artifact needed is Cluster API's go module appropriately tagged and pushed.
+
+These tests will be consumed by a provider just like any other go library dependency. They can be imported and provider custom types can be passed in and the e2es will be run for the given provider, assuming secrets are also managed
+
+The exact interface that a provider must satisfy is to be determined by what the tests demand. The plan for implementation at a high level is to build out reusable but pluggable testing components. For example, the framework will need a way to start a management cluster. We will use KIND initially and abstract the actual details away and expose an interface. If you want to use a different style of management cluster, say, a cloud-formation based cluster, then you will have to wrap up the cloud-formation code in some object and implement the testing framework’s exposed interface for the management cluster and use your object where the sample code uses the KIND-based management cluster struct. I expect but cannot guarantee that there will be interfaces for each provider but exact details are to be determined as the implementation gets more fleshed out.
```

Comment:
It sounds like there will be an interface for executing against CAPI APIs to setup / teardown clusters. Do we know yet if we need interfaces that query cloud providers directly to ensure resources are configured properly or are the e2es only validating what's available via CAPI resources or TBD?

Reply:
This is only testing what is available to CAPI resources. Cloud configuration is explicitly out of scope and should be tested by the providers directly.

The aim is high level behavioral tests. If any test seems too low-level please call it out. For example, we want to make sure clusters are creating a secret how cluster api expects. We can tell the cloud provider that is plugged in to create a cluster then we can use Cluster API primitives to go find the secret for that cluster and make sure it can contact the api server over TLS.